src/reflect/scala/reflect/internal/Types.scala

```diff
@@ -3131,13 +3131,53 @@ trait Types
        */
       def unifyFull(tpe: Type): Boolean = {
         def unifySpecific(tp: Type) = {
-          sameLength(typeArgs, tp.typeArgs) && {
+          if(sameLength(typeArgs, tp.typeArgs)) {
             val lhs = if (isLowerBound) tp.typeArgs else typeArgs
             val rhs = if (isLowerBound) typeArgs else tp.typeArgs
             // This is a higher-kinded type var with same arity as tp.
             // If so (see SI-7517), side effect: adds the type constructor itself as a bound.
             isSubArgs(lhs, rhs, params, AnyDepth) && { addBound(tp.typeConstructor); true }
-          }
+          } else if(settings.YhigherOrderUnification && typeArgs.lengthCompare(0) > 0 && compareLengths(typeArgs, tp.typeArgs) < 0) {
+            // Simple algorithm as suggested by Paul Chiusano in the comments on SI-2712
+            //
+            //   https://issues.scala-lang.org/browse/SI-2712?focusedCommentId=61270
+            //
+            // Treat the type constructor as curried and partially applied, we treat a prefix
+            // as constants and solve for the suffix. For the example in the ticket, unifying
+            // M[A] with Int => Int this unifies as,
+            //
+            //   M[t] = [t][Int => t]
+            //   A = Int
+            //
+            // A more "natural" unifier might be M[t] = [t][t => t]. There's lots of scope for
+            // experimenting with alternatives here.
+
+            def mkFreeSym(owner: Symbol, absSym: TypeSymbol): TypeSymbol = {
+              val name = currentFreshNameCreator.newName("Unify$")
+              val freeSym = owner.newTypeParameter(name)
+              val bounds = absSym.info.bounds
+              val info =
+                absSym.typeParams match {
+                  case Nil    => bounds
+                  case params => PolyType(params.map(sym => mkFreeSym(freeSym, sym.asType)), bounds)
+                }
+              freeSym setInfo info
+            }
```

Comment:
I think what you are after here is just `absSym.cloneSymbol`. That takes care of patching up references, such as type params that are bounded by other type params.

```
scala> def uniq[A](f : => A) = {settings.uniqid.value = true; try f finally settings.uniqid.value = false}
uniq: [A](f: => A)A

scala> trait T[M[A, X[_ <: A]]]
warning: there were two feature warnings; re-run with -feature for details
defined trait T

scala> val TClass = symbolOf[T[Any]]
TClass: $r.intp.global.TypeSymbol = trait T

scala> uniq(TClass.info.typeParams.head.defString)
res25: String = M#418429[A#418431, X#418432[_#418433 <: A#418431]]

scala> uniq(TClass.info.typeParams.head.cloneSymbol(newOwner = TClass).defString)
res26: String = M#444953[A#444954, X#444955[_#444957 <: A#444954]]
```